source/openpulse/openpulse/parser.py

```diff
@@ -0,0 +1,303 @@
+"""
+=============================
+Parser (``openpulse.parser``)
+=============================
+
+Tools for parsing OpenPulse programs into the :obj:`reference AST <openpulse.ast>`.
+
+The quick-start interface is simply to call ``openpulse.parse``:
+
+.. currentmodule:: openpulse
+.. autofunction:: openpulse.parse
+
+The rest of this module provides some lower-level internals of the parser.
+
+.. currentmodule:: openpulse.parser
+.. autoclass:: OpenPulseNodeVisitor
+"""
+
+# pylint: disable=wrong-import-order
+
+__all__ = [
+    "parse",
+    "OpenPulseNodeVisitor",
+]
+
+from contextlib import contextmanager
+from typing import List, Optional
+
+try:
+    from antlr4 import CommonTokenStream, InputStream, ParserRuleContext
+except ImportError as exc:
+    raise ImportError(
+        "Parsing is not available unless the [parser] extra is installed,"
+        " such as by 'pip install openqasm3[parser]'."
+    ) from exc
+
+import openpulse.ast as openpulse_ast
+from openqasm3.antlr.qasm3Parser import qasm3Parser
+from openqasm3 import ast
+from openqasm3.parser import (
+    span,
+    QASMNodeVisitor,
+    _raise_from_context,
+    parse as parse_qasm3,
+)
+from openqasm3.visitor import QASMVisitor
+
+from .antlr.openpulseLexer import openpulseLexer
+from .antlr.openpulseParser import openpulseParser
+from .antlr.openpulseParserVisitor import openpulseParserVisitor
+
+
+def parse(input_: str) -> ast.Program:
+    """
+    Parse a complete OpenPulse program from a string.
+
+    :param input_: A string containing a complete OpenQASM 3 program.
+    :return: A complete :obj:`~ast.Program` node.
+    """
+    qasm3_ast = parse_qasm3(input_)
+    CalParser().visit(qasm3_ast)
+    return qasm3_ast
+
+
+def parse_openpulse(input_: str, in_defcal: bool) -> openpulse_ast.CalibrationBlock:
+    lexer = openpulseLexer(InputStream(input_))
+    stream = CommonTokenStream(lexer)
+    parser = openpulseParser(stream)
+    tree = parser.calibrationBlock()
+    result = (
+        OpenPulseNodeVisitor(in_defcal).visitCalibrationBlock(tree)
+        if tree.children
+        else openpulse_ast.CalibrationBlock(body=[])
+    )
+    return result
```

Comment:
**question**: I don't quite understand when we'd hit the `else` statement. Is this when there's an empty `cal` statement?

Reply:
Exactly. Empty `cal` statement.